also introduce MarshalCSV for jsonenum-ed types

hchang-clypd opened the pull request:
https://jira.xandr-services.com/browse/CLYPD-41195

hchang-clypd committed:
also introduce MarshalCSV for jsonenum-ed types

APILLSBURY commented on template.go:
```diff
@@ -129,4 +129,9 @@ func (r {{$typename}}) Value() (driver.Value, error) {
 
 {{end}}
```
what's this {{end}}? Does your MarshalCSV func need to be before that?

hchang-clypd replied:
😬  good catch 472af33

APILLSBURY reviewed (changes requested):
one worry. Also should we make an UnmarshalCSV func while we're at it? Seems good to keep them in sync, though I don't imagine we'll want UnmarshalCSV often

hchang-clypd committed:
write MarshalCSV for every type needed

This was previously accidentally put outside of the template loop and
would've resulted in a var undefined error.

hchang-clypd committed:
also introduce UnmarshalCSV

hchang-clypd commented:
@APILLSBURY  yeah, I was hesitant to introduce the unmarshal because it's not always consistent across all the things we do. But upon more thinking I think this is generally safe to do, so 2da9b05

APILLSBURY commented on template.go:
```diff
@@ -132,6 +132,11 @@ func (r {{$typename}}) MarshalCSV() (string, error) {
 }
 
+// UnmarshalCSV is generated so {{$typename}} satisfies csv.Unmarshaler.
+func (r {{$typename}}) UnmarshalCSV(s string) (error) {
```
I think this needs to be a pointer receiver

hchang-clypd replied:
thanks. not having tests to rely on is scary 38a278e

APILLSBURY replied:
that's what code review is for :)

hchang-clypd committed:
UnmarshalCSV needs to be a pointer receiver

🤦‍♂️

enetland reviewed (commented):
looks reasonable. Are we going to run `go generate` across the whole repo to introduce these for existing types? might be worth doing on at least one type to provide a good path to test before merging

hchang-clypd commented:
@enetland  yes, my plan after merging this is
* rebuild and update the binary on clypd-engineering s3 bucket
* regenerate all our jsonnum things with this enhancement

enetland reviewed (approved):
ah right i forgot we have to rebuild etc. Plan sounds good, though maybe still a good idea to build locally and try it out on one type first to avoid having to redo the whole flow if we missed something

hchang-clypd commented:
local testing steps:
* `$ GO111MODULE=on go get github.com/clyphub/jsonenums@38a278e` to install the binary
* `$ cp bin/jsonenums bin/third_party/`
* `$ cd src/clypd/models/tvplatform && go generate`

Which yielded the following diff:
```
hc072p@Herberts-MacBook-Pro go % git diff                    
diff --git a/go/src/clypd/models/tvplatform/goaltype_jsonenums.go b/go/src/clypd/models/tvplatform/goaltype_jsonenums.go
index cf39a6bcf3..3a21337146 100644
--- a/go/src/clypd/models/tvplatform/goaltype_jsonenums.go
+++ b/go/src/clypd/models/tvplatform/goaltype_jsonenums.go
@@ -117,3 +117,13 @@ func (r *GoalType) Scan(i interface{}) error {
 func (r GoalType) Value() (driver.Value, error) {
        return r.getString()
 }
+
+// MarshalCSV is generated so GoalType satisfies csv.Marshaler.
+func (r GoalType) MarshalCSV() (string, error) {
+       return r.getString()
+}
+
+// UnmarshalCSV is generated so GoalType satisfies csv.Unmarshaler.
+func (r *GoalType) UnmarshalCSV(s string) error {
+       return r.setValue(s)
+}
diff --git a/go/src/clypd/models/tvplatform/rfpstatus_jsonenums.go b/go/src/clypd/models/tvplatform/rfpstatus_jsonenums.go
index 6771ed1c23..7ebb101f91 100644
--- a/go/src/clypd/models/tvplatform/rfpstatus_jsonenums.go
+++ b/go/src/clypd/models/tvplatform/rfpstatus_jsonenums.go
@@ -132,3 +132,13 @@ func (r *RFPStatus) Scan(i interface{}) error {
 func (r RFPStatus) Value() (driver.Value, error) {
        return r.getString()
 }
+
+// MarshalCSV is generated so RFPStatus satisfies csv.Marshaler.
+func (r RFPStatus) MarshalCSV() (string, error) {
+       return r.getString()
+}
+
+// UnmarshalCSV is generated so RFPStatus satisfies csv.Unmarshaler.
+func (r *RFPStatus) UnmarshalCSV(s string) error {
+       return r.setValue(s)
+}
```

I then rebuilt `tvpddeliverycmd`, which is a test binary to create sample CSV data for the TVDP dashboard data stream:
`$ make bin/tvpddeliverytestcmd`, and generated a sample file with no issues and the column values now serialize as I expect
